### ISSUE TITLE
[debian package] installing package maintainers config results in massive logouts and key loss.

### ISSUE THREAD

**Issue by gerg5c42g542g2c54g52c** [matrix-org/synapse#4658]
When updating synapse, you're asked to use package maintainers homeserver.yaml, if you agree, synapse will automatically restart and everyone on the HS will be logged out and their keys be lost. There should either be a prevention measure or a clear warning (preferably requiring another confirmation) when doing so. The results are disastrous and it's easy to make a mistake when you're tired or inexperienced.

**Comment by hawkowl** [comment]
@gerg5c42g542g2c54g52c is this the matrix.org debian package or the Debian-backports one?

**Comment by gerg5c42g542g2c54g52c** [comment]
It's the matrix.org package.

**Comment by richvdh** [comment]
This is presumably only true if you have changed `macaroon_secret_key` or `registration_shared_secret`?

**Comment by gerg5c42g542g2c54g52c** [comment]
I changed `registration_shared_secret`, but I think it has more to do with the fact that the postgresql db isn't connected to synapse anymore (the default config uses sqlite) or default federation endpoint configuration being different. It's not a real bug, more like a poor design, it's simply too easy to shoot your leg.

**Comment by richvdh** [comment]
I'm not really sure what more we can do if you choose to overwrite your config so that synapse can't find its database any more?

**Comment by gerg5c42g542g2c54g52c** [comment]
Warn of the consequences; don't automatically restart synapse and warn about this; provide the new config, but disabled fe. as homeserver.yaml.dpkg-new; including a sufficient mention of this in docs, something like 'if you want to revert to or use a clean config do this, but note that you'll have downtime', perhaps provide a current default config somewhere to download or info on how to extract it?

I see that the default behavior is moving the original config to homeserver.yaml.dpkg-old and putting the new one as the default, which is the opposite to what people want. That might be a debian thing though.

**Comment by richvdh** [comment]
> Warn of the consequences; 

replacing the configuration is done by dpkg - we don't have an opportunity to add a warning. 

> don't automatically restart synapse and warn about this; 

Not restarting synapse on package upgrade would be more surprising to most users; we used to get frequent bug reports about it until this was fixed.

> provide the new config, but disabled fe. as homeserver.yaml.dpkg-new; 

That would mean that synapse wouldn't work out-of-the-box, which again would be more surprising to most users.

> including a sufficient mention of this in docs, something like 'if you want to revert to or use a clean config do this, but note that you'll have downtime and users logged out'; 

I'm a bit unconvinced that anyone would read any documentation like this.

> perhaps provide a current default config somewhere to download 

This is https://github.com/matrix-org/synapse/issues/1433, and certainly something I want to address.

> or info on how to extract it or just say 'don't do this!'?

I don't really know what you mean by this.

> I see that the default behavior is moving the original config to homeserver.yaml.dpkg-old and putting the new one as the default, which is the opposite to what people want. That might be a debian thing though.

It is a debian thing.

Basically, on this, I understand your pain, but it's hard to do much about it.

One thing that I think might help is to suggest that people put config overrides in a file other than `homeserver.yaml`, so that they are preserved on upgrade.

**Comment by neilisfragile** [comment]
Okay this is bad situation and I'm sorry you lost your keys - the impact of the log outs will be reduced by features like https://github.com/matrix-org/matrix-doc/blob/55ebf7dc3275822ad0126546baef4b1361418baf/proposals/1466-soft-logout.md

In terms of concrete todos #1433 seems to cover it - I don't think there is much else we can do here. Closing this issue.

**Comment by gerg5c42g542g2c54g52c** [comment]
> > provide the new config, but disabled fe. as homeserver.yaml.dpkg-new;
> 
> That would mean that synapse wouldn't work out-of-the-box, which again would be more surprising to most users.

I meant doing this only if upgrading.

>> or info on how to extract it or just say 'don't do this!'?
>
> I don't really know what you mean by this.

Extract the config from the .deb package.

> One thing that I think might help is to suggest that people put config overrides in a file other than `homeserver.yaml`, so that they are preserved on upgrade.

That's a very good idea. This would fix the problem and should also makes things cleaner and easier to maintain. Docs should recommend doing any config changes in a `homeserver.yaml` file located in `/etc/matrix-synapse/conf.d`.
